Bring the design section of the handbook up to date

kjellr opened the pull request:
I've been working on a rewrite + restructuring of the Design section of the Gutenberg handbook. The goals are:

- To get everything up to date. A number of our current descriptions + screenshots do not reflect the current state of Gutenberg.
- Replace most of our screenshots with more generic-looking "blueprints" so that they act more as wireframes than 1:1 design representations. This will help make sure they aren't out of date the moment something in the UI moves. Alongside that, I aimed to include more annotated visuals where I felt they'd be helpful.
- Include mobile representations of the UI alongside all of the desktop screens.
- Turn the interface blueprints into more of a comprehensive "Design Patterns" section. Add missing elements like the Block Library. 
- Update the names of interface elements to be more in line with our commonly used titles (via this [reference](https://github.com/karmatosed/talk-gutenberg-anatomyofablock/blob/master/anatomyofablock.pdf))
- Update the sketch file link to point to [SketchPress](https://github.com/10up/SketchPress). The designers are in the process of [figuring out the best approach](https://make.wordpress.org/design/2018/06/27/design-meeting-notes-for-june-27-2018/#comment-25587) for our Sketch files, but in the meantime we all love the SketchPress files. They're also updated more frequently than the ones we used to link to. 
- Also, I aimed to generally tidy up the copy just a little bit throughout.

This is a fairly comprehensive rework. The vast majority of the content that was there still exists, but has been cleaned up and moved around. 

Because this adds a lot of new content, I propose we break the "Design Principles" page out of the "Reference" section, and into its own top-level "Design" item, with 3 subpages. I've done that in this PR, but I'm 100% happy to change it. The new structure of the design content is as follows: 

- [Design](https://github.com/WordPress/gutenberg/blob/update/design-handbook/docs/design.md) (Overview, high-level principles + vision)
	- [Patterns](https://github.com/WordPress/gutenberg/blob/update/design-handbook/docs/design/design-patterns.md) (Overview of commonly used design elements + structure)
	- [Block Design](https://github.com/WordPress/gutenberg/blob/update/design-handbook/docs/design/block-design.md) (Best practices + examples for designing blocks)
	- [Resources](https://github.com/WordPress/gutenberg/blob/update/design-handbook/docs/design/design-resources.md) (Links to the SketchPress library)

_(For those looking to review, I'd suggest reading through these pages using the links above.)_

Looking forward to thoughts  + feedback.

**Still to do:**

- [x] This could definitely use a copy review. 
- [x] I'm actually not sure how the `manifest.json` works. 😄 I made a quick update in this PR, but I'd appreciate someone with more knowledge stepping in to verify that it's set up correctly.

kjellr committed:
Adding revised design handbook strucutre + content

kjellr committed:
Adding revised design handbook strucutre + content

kjellr committed:
Removing old Design doc from the Reference section.

kjellr committed:
Removing old Design doc from the Reference section.

kjellr committed:
(Attempting to) Add new design doc pages/structure to manifest.json

kjellr committed:
(Attempting to) Add new design doc pages/structure to manifest.json

kjellr committed:
Properly closing the code block in the theme example.

kjellr committed:
Properly closing the code block in the theme example.

kjellr committed:
Properly opening the code block in the theme example.

kjellr committed:
Properly opening the code block in the theme example.

karmatosed reviewed (approved):
I agree we should get a copy review but to me these are all great changes. This makes a lot of sense to me and thanks for taking on this big task.

sarahmonster committed:
Improve pre-publish panel styling (#7879)

* Use clearer wording for pre-publish panel prompt; fixes #7873.
* Style pre-publish panel more consistently with settings sidebar; fixes #7878.
* Switch to an all-white background for prepublish panel.

vindl committed:
Packages: update api-fetch package documentation (#8730)

Add a minor fix to the api-fetch middleware documentation
related to overriding the API root URL.

earnjam committed:
Prevent Travis from running when changes are only made to .md files (#8728)

This comes from [a discussion](https://wordpress.slack.com/archives/C02QB2JS7/p1533720310000209) with @gziolo in #core-editor. Basically we can skip the full Travis build if the commit range being tested only contains changes to `.md` files. 

Modeled after: https://github.com/facebook/react/pull/2000

In my testing, this shortens the Travis run time on each item in the test list from ~4-6 minutes down to ~1 minute or less for docs only changes. Should help speed up merging non-code PRs. 

Here are some test builds I did from my fork:
[Only change .md file](https://travis-ci.org/earnjam/gutenberg/builds/413732648) - aborts early
[Only change .js file](https://travis-ci.org/earnjam/gutenberg/builds/413733458) - runs normally
[Changes both .md and .js files](https://travis-ci.org/earnjam/gutenberg/builds/413733882) - runs normally

afercia committed:
Additional keyboaard shortcuts for navigateRegions. (#8005)

gziolo commented on docs/manifest.json:
```diff
@@ -83,6 +83,30 @@
 		"markdown_source": "https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/extensibility/autocomplete.md",
 		"parent": "extensibility"
 	},
+	{
```
This file is autogenerated, you need to update `root-manifest.json` file instead.

gziolo replied:
Ah, once you update the proper file, you will need to run `npm run docs:build` to regenerate this file 🙃

kjellr replied:
😄 Ah, that does it! Thanks @gziolo. Updated in https://github.com/WordPress/gutenberg/pull/8726/commits/03905f2735e9d3fd7a2cd926a6bdce7cee95612b

desrosj committed:
Manually print inline metabox initialization JavaScript when the `wp-edit-post` script is printed in the `<head>`.

Fixes #6963.

jasmussen committed:
Sort strikethrough after link (#8755)

This is a tiny PR to sort the "Strikethrough" button after the link.

One of the... perhaps this is unwritten and should be written down somewhere... one of the design principles of the block toolbar is that it starts with block related commands (like switcher), then _block level_ (i.e. actions that apply to the whole block, like textalign applies for the whole paragraph), and finally _inline level actions_ last. In the latter, it is sorted by presumed order of importance, and in this case, I would like to argue that Link is more important than Strikethrough.

One side effect of this is that one day when we can make the toolbar better handle nested contexts or small screens, we can "pop off" buttons into an overflow menu as the toolbar has to scale. If/when that happens, it's important that buttons are also sorted in order of importance.

noisysocks committed:
Fix 'Prevent duplicate DotTips from appearing' feedback (#8741)

* Typo: s/dispathed/dispatched/

* NUX: Simplify REGISTER_TIP_INSTANCE reducer

koke committed:
Expose redux-routine to React Native (#8761)

aduth committed:
Testing: Trigger E2E test failure on console logging (#8721)

* Testing: Extract post trashing to function

* Testing: Remove all page events on teardown

* Testing: Trigger E2E test failure on console logging

ntwb committed:
Update Lerna to 3.0.0-rc.0 (#8759)

kjellr committed:
Revert "(Attempting to) Add new design doc pages/structure to manifest.json"

This reverts commit 92619396c57d08a74b4e026044f0ee816c1d389b.

kjellr committed:
Revert "(Attempting to) Add new design doc pages/structure to manifest.json"

This reverts commit 92619396c57d08a74b4e026044f0ee816c1d389b.

sarahmonster commented:
❤️ the diagrams here especially—so helpful to start developing a shared language for what we're talking about.

The new hierarchy is great, but it repeats the word "design" a whole lot:
> Design (Design Patterns, Block Design, Design Resources)

What about instead:
> Design (Patterns, Block Design, Resources)

I have some nitpicky copy suggestions. Would you prefer comments or commits? 🙂

kjellr committed:
Rename pages to prevent the repeated use of "Design"

"Design Patterns" > "Patterns"
"Design Resources" > "Resources"

These all live under the "Design" folder anyway, so I think it's fine to remove that part of the title.

kjellr committed:
Rename pages to prevent the repeated use of "Design"

"Design Patterns" > "Patterns"
"Design Resources" > "Resources"

These all live under the "Design" folder anyway, so I think it's fine to remove that part of the title.

kjellr committed:
Add new pages/structure to root-manifest.json

kjellr committed:
Add new pages/structure to root-manifest.json

kjellr commented:
@sarahmonster:

> What about instead:
> Design (Patterns, Block Design, Resources)

Sounds good to me. Updated in https://github.com/WordPress/gutenberg/pull/8726/commits/e881c559a42d7fe26cf91f6b75c4eb5d91e70122

> I have some nitpicky copy suggestions. Would you prefer comments or commits? 🙂

Feel free to commit!

gziolo commented on docs/root-manifest.json:
```diff
@@ -83,6 +83,30 @@
+	},
+	{
+		"title": "Resources",
+		"slug": "resources",
```
We already have slug resources. I'm not sure if this won't get into conflict when processed by the script which maintains Gutenberg handbook. I would recommend the following changes to ensure it works properly on wordpress.org and we can link files in both GitHub and handbook:

File renamed to `design-resources.md`.

```json
"slug": "design-resources",
"markdown_source": "https:\/\/raw.githubusercontent.com\/WordPress\/gutenberg\/master\/docs\/design\/design-resources.md",
```

gziolo commented on docs/root-manifest.json:
```diff
@@ -83,6 +83,30 @@
+	},
+	{
+		"title": "Patterns",
+		"slug": "patterns",
```
Similar to `resources` - patterns is very short for the slug name, I think we should we do the following:

File renamed to `design-patterns.md`.

```json
"slug": "design-patterns",
"markdown_source": "https:\/\/raw.githubusercontent.com\/WordPress\/gutenberg\/master\/docs\/design\/design-patterns.md",
```

kjellr committed:
Update filenames + slugs.

Changing "patterns" to "design-patterns", and "resources" to "design-resources" to prevent conflict with other slugs we're already using.

kjellr committed:
Update filenames + slugs.

Changing "patterns" to "design-patterns", and "resources" to "design-resources" to prevent conflict with other slugs we're already using.

ntwb committed:
chore(release): publish

 - @wordpress/a11y@1.1.3
 - @wordpress/api-fetch@1.1.1
 - @wordpress/autop@1.1.3
 - @wordpress/babel-plugin-import-jsx-pragma@1.0.3
 - @wordpress/babel-plugin-makepot@2.0.3
 - @wordpress/babel-preset-default@2.0.4
 - @wordpress/blob@1.0.3
 - @wordpress/blocks@2.0.0
 - @wordpress/components@2.0.0
 - @wordpress/compose@1.0.3
 - @wordpress/core-data@1.1.0
 - @wordpress/custom-templated-path-webpack-plugin@1.1.4
 - @wordpress/data@1.2.0
 - @wordpress/date@1.0.3
 - @wordpress/deprecated@1.0.3
 - @wordpress/dom-ready@1.2.0
 - @wordpress/dom@1.0.3
 - @wordpress/editor@2.0.0
 - @wordpress/element@1.0.3
 - @wordpress/hooks@1.3.3
 - @wordpress/html-entities@1.0.3
 - @wordpress/i18n@1.2.3
 - @wordpress/is-shallow-equal@1.1.3
 - @wordpress/jest-console@2.0.3
 - @wordpress/jest-preset-default@2.0.3
 - @wordpress/keycodes@1.0.3
 - @wordpress/library-export-default-webpack-plugin@1.0.3
 - @wordpress/nux@1.1.0
 - @wordpress/plugins@1.0.3
 - @wordpress/postcss-themes@1.0.3
 - @wordpress/redux-routine@1.0.0
 - @wordpress/scripts@2.1.0
 - @wordpress/shortcode@1.0.3
 - @wordpress/url@1.2.3
 - @wordpress/viewport@1.0.3
 - @wordpress/wordcount@1.1.3

gziolo reviewed (approved):
Thanks for improving all the stuff related to the manifest 💯 Looks good for me on non-design related angle :)

diegoreymendez committed:
Publishes unknown-type handler methods for mobile. (#8688)

dimadin committed:
Capital P, dangit (#8772)

sarahmonster committed:
Wee copyedits to main design page.

sarahmonster committed:
Wee copyedits to main design page.

kjellr commented:
@chrisvanpatten Very cool! I could definitely see some of that content merging in here. 👍 After this merges in, let's start a new PR to discuss.

gziolo committed:
Element: Improve validation for attribute names in serializer (#8775)

aduth committed:
RichText: Fix empty value delete behaviors (#8735)

* Testing: Wait for RichText initialization

* RichText: Fix empty value delete behaviors

* RichText: Handle remove only on backspace

* docs: Very minor comment readability tweak

mcsf committed:
i18n: Add context to occurrences of 'More' string (#8766)

aduth committed:
Update plugin version to 3.5 (#8784)

Commit:
Localization & Spelling fixes (#8795)

Fixes #8794.

talldan committed:
Add new keyboard shortcuts for block settings menu (#8279)

Adds new keyboard shortcuts for the block settings menu:

Cmd+Shift+D / Ctrl+Shift+D - duplicate block
Cmd+shift+, / Ctrl+shift+, - show/hide sidebar

* Add duplicate shortcut to block settings menu

- Move onDuplicate dispatch from BlockDuplicateButton into a new HOC called
`withBlockSettingsActions`
- Add `shortcuts` as a prop in `withBlockSettingsActions` that is passed to
its wrapped component
- Pass onDuplicate and shortcut through to BlockDuplicateButton
- Add new BlockSettingsKeyboardShortcuts component for handling shortcut
functionality

Make BlockDuplicateButton a dumb component - move HOC logic to withBlockSettingsActions

Add keyboard shortcut for block removal (Cmd/Special/Windows + DEL)

- Move block remove action dispatch from BlockRemoveButton up to withBlockSettingsActions
- Make BlockRemoveButton a dumb component
- Add keyboard shortcut for block removal to BlockSettingsKeyboardShortcuts
- Pass onRemove into KeyboardShortcuts and BlockRemoveButton

Add shortcut for toggling the sidebar

- Add shortcut keys to list of keyboard shortcuts for edit-post
- Add open/close dispatchers to EditorModeKeyboardShortcuts and hook up
- Make BlockInspectorButton display the keyboard shortcut

MenuItem should use aria-checked attribute instead of aria-pressed

See example implementation of a MenuBar here for details:
https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-2/menubar-2.html

Provide a default role for MenuItem component of menuitem

Still allows the role to be overriden for other menuitem types
(i.e. menuitemcheckbox, menuitemradio), but defaults to menuitem.

Also makes the provision of isSelected prop optional since the resulting
aria-checked role should not be used with a menu-item.

Update usages of MenuItem that should not use the default role

Switch menu items in block settings menu that have a shortcut to use MenuItem

Use the MenuItem component which correctly renders the shortcut. The need for
wrapping components BlockDuplicateButton and BlockRemoveButton is minimal, so
remove them.

Remove role props, since MenuItem implements a default

Make all block settings menu items use MenuItem component instead of IconButton

Spread additional props for MenuItem onto rendered component

This allows MenuItem to support additional props like `disabled`

Remove stroke being applied to MenuItem icons

Most icons don't have stroke - this style rule was adding an additional stroke
to SVG icons, making them appear chunky and blurry.

Remove use of !important

Fix lack of left margin on menu items with icons

Fix potential access of undefined property

When removing the block, there's momentarily a null item in the blocks array

Render block settings when typing, but make it hidden

This ensures BlockSettingsKeyboardShortcuts is rendered and shortcuts
work while actively typing

* Rename shortcut keycodes to raw/display to match block settings menu

* displayShortcut removes `+` symbol between shortcuts when the provided key is any single character

* displayShortcut capitalizes provided key (e.g. Del instead of DEL)

* Avoid placing a + between ⌘ and key entirely

* Change duplicate shortcut to Cmd/Ctrl shift D

* Change toggleSidebar shortcut to `cmd/ctrl + shift + ,` to avoid conflict with Chrome settings shortcut

* When toggling the sidebar using the keyboard, open the correct tab based on whether block(s) are selected

* Add keyboard shortcuts to tooltips for closing and opening the settings sidebar

* Make use of stopImmediatePropagation less zealous

Code is supposed to prevent further handling of keyboard events when a block is
deleted, but doesn't consider cases where a block is not deleted:

- If the user hits delete and the block is kept, but content from another block
merged in
- If the user hits delete, but the block has content and is thus not deleted

* Add modifier key exceptions to prevent deletion or merge of blocks

* Make block deletion shortcut Cmd+Opt+Backspace or Cmd+Opt+Delete

- For Windows that's Ctrl+Alt+Backspace or Ctrl+Alt+Delete

* Add further clauses where event should return early

Iron out inconsistent behaviour with use of backspace/delete key when it can
result in deletion or merging of blocks. When specific modifiers are pressed
the event should not have any effect. Here, the use of preventDefault and
stopImmediatePropagation were blocking further binding of keyboard shortcuts
and preventing browser/system level keyboard shortcuts.

* Make keyboard shortcut for delete Cmd+Alt+Delete

* Abbreviate display shortcut to prevent line wrapping in block settings menu

* Merge `withBlockSettingsActions` into `BlockSettingsMenu`

- BlockSettingsMenu was the only consumer of withBlockSettings
- shortcuts object is now not injected into props

* Remove BlockSettingsKeyboardShortcuts component, which had no logic

* Use preventDefault to prevent any additional browser/os shortcuts and add comments

* Use object method shorthand

* Add !important back in, with clarifying comment

* Use lodash noop over noop arrow function

* Destructure dispatching functions

* Refine exceptions to RichText Backspace/Delete handling

* Remove the Remove Block shortcut

* Revert changes to RichText back to master version

* Fix unit tests (stylelint)

talldan committed:
Add new keyboard shortcuts help modal (#8316)

Pressing ctl+option+h (shift+alt+h on windows/linux) displays a modal detailing keyboard shortcuts.

* Create a simple modal for showing keyboard shortcuts that displays when pressing cmd + /

* Add more shortcuts and place them in sections with a title and description

* Add styles for shortcut modal

* Add missing shortctuts

* Remove unecessary shortcut section descriptions

* Display shortcut keys in a kbd element

* Styling changes

- Position shortcut to the right of description
- Adjust margin/padding
- Make section titles default font size

* Regroup shortcuts

* Fix invalid style rule

* Change copy for slash inserter shortcut

* Map over config to produce ShortcutSections

* Change where keyboard shortcut help modal is rendered

Move modal so that it's rendered in edit-post/layout component. This ensures
the modal is available in both the text and visual editor, and the shortcut
for launching the modal can be used on first load of the editor.

* Move keyboard-shortcut-help-modal to edit-post from editor package

* Add some border between rows in the keyboard shortcuts list

* Add reduxy things for opening and closing a modal

* Use actions to open/close modal and selector to determine current state

* Migrate the activeModal state from preferences to normal state so that its value is not persisted

* Add a menu item tot he more menu for toggling keyboard shortcut help

* Modify some shortcuts and description text

* Add snapshot tests for help modal

* Destructure props

* Modify header styling

* Rename class name to reflect component is now in edit-post, and reorder shortcuts

* Display shortcut for more menu menu item

* Disable WP_Help TinyMCE command for classic block

* Change keyboard shortcut for triggering help to access + h, the same as the classic editor

* Add utility to keycodes for returning key combinations in an array

* Avoid ugly regular expression to split key combination into array

Using newly introduced displayShortcutList function results in cleaner, better tested code

* Add e2e test for shortcut help modal

* Vertically center shortcut text (if adjacent label wraps onto multiple lines)

* Close more menu when keyboard shortcuts modal is opened using the menu item

* Use castArray over lengthy ternary

* Styling linter fixes

* JSDoc fixes

* Group activeModal unit tests in a describe

* Add test for initial activeModal state

* Fix incorrect state used in test

* Update list of key shortcuts for the help modal

* Remove unecessary return

* Make KeyboardShortcutsHelpMenuItem only open the help modal

jasmussen committed:
Pad menus (#8758)

* Pad menus

This adds padding to the block more menu. At first glance this will look like much, but in fact it simply makes it the same padding as that of every other popover.

* Sort again.

koke committed:
Port heading block to mobile (#8774)

* Adds tagName prop to RichText

Instead of forcing a paragraph, make the wrapper tag logic more generic
and allow for customization

* Adds React Native implementation of heading block

* Fix lint issues

* Remove RichText tagName default value

* Remove unused (for now) class name props

youknowriad committed:
Audio Block: Avoid scrollbars when shown full-width (#8820)

jasmussen committed:
Make active pill button look pressed. (#8819)

* Make active pill button look pressed.

This is a small change that makes the activated pill button look pressed. Right now it just looks like a button, and while it is, it's an _active_ button.

* Fix typo.

sarahmonster committed:
Minor copyedits to block design page.

sarahmonster committed:
Minor copyedits to block design page.

sarahmonster committed:
Copyedit design resources page.

sarahmonster committed:
Copyedit design resources page.

notnownikki committed:
Allow access to the WordPress installation if DOCKER_ENV=localwpdev (#8818)

youknowriad committed:
Extract Core Blocks to the npm packages folder (#8287)

* Extract Core Blocks to the npm packages folder

* Core blocks: Adjustment after rebase with master

* Build: Add support for all scss files in build-style directory

* Build: Fix output path for styles generated for packages

* Fix sass files after rebase

* Fix watching SASS files

* Rename core-blocks block-library

* Fix linting issues

* Update docs

* Deprecate Core Blocks

* Update index.js

* Update client-assets.php

* Update client-assets.php

sarahmonster committed:
Tiny copyedit for clarity on the design patterns page.

sarahmonster committed:
Tiny copyedit for clarity on the design patterns page.

sarahmonster reviewed (approved):
I give this an A+. 

(Also I did some wee copyedits all over, please feel free to tweak/untweak as needed!)

kjellr committed:
Copyedits to the Design Principles & Vision page

props @michelleweber.

kjellr committed:
Copyedits to the Design Principles & Vision page

props @michelleweber.

kjellr committed:
Merge branch 'update/design-handbook' of https://github.com/WordPress/gutenberg into update/design-handbook

kjellr committed:
More copyedits.

props @michelleweber

kjellr committed:
More copyedits.

props @michelleweber

kjellr committed:
Merge branch 'update/design-handbook' of https://github.com/WordPress/gutenberg into update/design-handbook

kjellr commented:
^ Rebased to hopefully get that check to pass...

mizejewski commented:
@kjellr this is an **outstanding** re-work 👏 
I just read with fresh eyes the existing Reference > Design Principles and your proposed edits, and your version really enhances clarity. 

I like the shift to more generic-looking "blueprints" for the reasons you mention, however, it can get a little abstract (e.g. the explanation of a block chip in the Patterns section.) It might be helpful to mention/link to the more concrete examples in the Block Design section.

gziolo commented:
There are 548 files changes, I don't think it's right. I'm seeing unrelated commits on the list. Something went wrong when merging changes with the master branch.

sarahmonster commented:
We merged these in with a clean branch in #8924, so unless I missed anything, all these changes have been taken care of and merged into master now.

kjellr commented:
@mizejewski, thanks for taking a look. (and @sarahmonster thanks for following through with the cleanup + merge!)

> It can get a little abstract (e.g. the explanation of a block chip in the Patterns section.) It might be helpful to mention/link to the more concrete examples in the Block Design section.

I think we could definitely expand some of the more potentially uncommon items (like block chips, which are based on ["Action Chips" in Material Design](https://material.io/design/components/chips.html#action-chips)), and also include some more explicit visual examples of the block design section. We'll aim to work that in to the next update.
